Fix miri validation errors through now stricter provenance

ten3roberts opened the pull request:
Miri has become even stricter with regards to pointer borrowing and stacked borrows, especially with the now default *full-field-retagging*

ten3roberts committed:
fix(miri): box transmute and invalid references

The general causes for the miri invalidation is the prevelant use of
`Box` and its references to `ErrorImpl<()>`.

`mem::transmute` does not preserve the tag stack for transmuting the
boxes.

Additionally, having references to `ErrorImpl<()>` which has a different
layout than the allocation or `ErrorImpl<E>` for some unknown `E`. This
causes the new "untyped" reference to now have a provenance that
includes the size of E and thus is outside the provenance.

ten3roberts committed:
fix(miri): downcast_mut using `&mut _ => *const _ => *mut`

ten3roberts committed:
fix(miri): stub file reading

ten3roberts commented:
Fixes #59

ten3roberts commented:
Remaining:
- [ ] Fullfill MSRV requirements
- [ ] Test miri in CI
- [ ] Don't use invalid references to get a ref to `handler` in `ErrorImpl`

ten3roberts committed:
fix(miri): don't construct temp references of shunk provenance

ten3roberts committed:
ci: miri

ten3roberts commented:
Miri now passes

The only thing which does not pass is that some of our dependencies (mainly once_cell) does not fullfill our MSRV requirement.

We need to either bump our MSRV or lock once_cell, though I'dprefer that call and discussion to go in a separate PR to limit scope.

@yaahc let me know when you have some time to look through this. Don't worry about being too pedantic; I am here to learn :blush:

yaahc commented on src/error.rs:
```diff
@@ -469,65 +504,66 @@ impl Deref for Report {
+unsafe fn object_drop<E>(e: OwnedPtr<ErrorImpl<()>>) {
     // Cast back to ErrorImpl<E> so that the allocator receives the correct
     // Layout to deallocate the Box's memory.
     // Note: This must not use `mem::transmute` because it tries to reborrow the `Unique`
     //   contained in `Box`, which must not be done. In practice this probably won't make any
     //   difference by now, but technically it's unsound.
     //   see: https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/stacked-borrows.md
```
Is this still accurate after these changes?

yaahc reviewed (commented):
It looks great. I didn't spend too much time trying to audit the correctness of the unsafe; I'm going to trust MIRI and you on that one. 

As for the question about once_cell and MSRV, I have no objections to resolving that separately, though I prefer resolving that first, landing that, and then rebasing that into this PR before merging. However, I am open to other plans if you have a different one in mind.

ten3roberts committed:
fix: `unsafe_op_in_unsafe_fn`

ten3roberts committed:
chore!: bump MSRV

ten3roberts commented on src/error.rs:
```diff
@@ -191,20 +198,23 @@ impl Report {
+        // let erased = mem::transmute::<Box<ErrorImpl<E>>, Box<ErrorImpl<()>>>(inner);
+        let ptr = OwnedPtr::<ErrorImpl<E>>::new(inner);
+
+        // Safety: the type
+        let ptr = ptr.cast::<ErrorImpl<()>>();
+        Report { inner: ptr }
```
Yes, the surrounding function is unsafe to *call* as it relies on the passed vtable being from the type `E`

ten3roberts committed:
chore: address PR comments

ten3roberts commented:
Should we merge this and tackle the failing test that looks like something entirely different in another PR, or should we solve it here?

I suppose this goes in general, do you prefer larger PRs which implement things in full, or smaller atomic PRs; as one would do on trunk based development?

yaahc commented:
> Should we merge this and tackle the failing test that looks like something entirely different in another PR, or should we solve it here?
> 
> I suppose this goes in general, do you prefer larger PRs which implement things in full, or smaller atomic PRs; as one would do on trunk based development?

I don't have a strong preference. I don't love having CI be broken on main but it's been a while since we were in that state. I guess I'd probably lean towards atomic PRs.

How about you, which do you prefer?

ten3roberts commented:
I'd say I am slightly towards somewhat bigger PRs where a feature is working for the most part before being merged in, rather than merging bit by bit and feature flagging it until it is done.

This way, you get a single place to refer to and have the diff gathered and duplicated (if you end up changing the same lines again in another PR)

I don't prefer broken CIs either, and I will see what I can do with the remaining checks 😊

ten3roberts committed:
fix: ci workflow names

ten3roberts committed:
chore: raise msrv to 1.65 (addr2line)

ten3roberts committed:
chore: revert distinctive CI names due to branch protection rules

The new names, such as `Test Platform Matrix` which do make it easier to
see which jobs failed, rather than msrv, test, and miri all being called
`Test Suite`, the in-place branch protection rules wait forever until
the now non-existent `Test Suite` passes

ten3roberts commented:
@yaahc CI seems to work now. Is it alright by you to merge this now?
